### ISSUE TITLE
When will the Multi-arched gcr.io/istio-testing/build-tools Image be Available

### ISSUE THREAD

**Issue by TrevorTaoARM** [istio/istio#35176]
(This is used to request new product features, please visit <https://discuss.istio.io> for questions on using Istio)

**Describe the feature request**

We know that most Istio related images (docker.proxyv2, docker.pilot, docker.app...) are built based on the gcr.io/istio-testing/build-tools:xxx image, and its Dockerfile in the istio/tools supports both amd64 and arm64 building now.

But now it seems the gcr.io/istio-testing/build-tools image still only supports amd64 arch. So I would like to ask when will the multi-arched ( amd64, arm64,...) enabled built-tools image would be available to us which would greatly facilitate our building and development of Istio on arm64 platform.

**Describe alternatives you've considered**

**Affected product area (please put an X in all that apply)**

[ ] Docs
[x] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[x] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

**Comment by howardjohn** [comment]
Likely not until we have ARM nodes in our CI, and all of our CI dependencies support ARM, and everything building+testing on ARM.  Until then we can only use qemu emulation which takes a very long time for the build-tools image.

**Comment by TrevorTaoARM** [comment]
Is there anything needed or now missing if we prefer to have a ARM CI? No arm server environment now?

**Comment by howardjohn** [comment]
I am not sure if our CI system (prow) entirely supports ARM yet either. It
might. https://github.com/kubernetes/test-infra/issues/16588

If that is done then we just need the servers + someone to invest some time
in wiring everything up

**Comment by howardjohn** [comment]
I think we also need to setup some tooling such that in postsubmit, we trigger two prowjobs/pods: one on amd64 and one on arm64. These build their respective images (ie build-tools-arm64)

Once both are built, they should be merged into a single multi-arch image (build-tools).

I am not actually sure how to do this. Typically with buildx you build multiple architectures at once, using emulation or external builders (kubernetes). Emulation is not viable for our builds, and kubernetes is not preferred - we do not want to give pods highly privileged access to create kubernetes pods. Since prow does not have DAG pipelines, we cannot trivially run action after both jobs complete to manually merge them.

On istio/proxy side its a bit easier since we don't need multi-arch image

**Comment by TrevorTaoARM** [comment]
> 

@howardjohn Thanks. 
I agree with the way that we trigger 2 prow jobs/pods on both platforms after postsubmit, and the merge problem could be considered further.  

One possible sequences to enable multi-arch images that I can imagine is listed below, which may not be so feasible and deeply considered.

1.1 Trigger building amd64 images on amd64 platform when postsubmit.
1.2 Trigger building arm64 images on arm64 platform when postsubmit.
2.1. Tag the amd64 images as HUB/{image_name}: {image_tag}-ARCH
2.2. Pushing 2 amd64 images to HUB: (now ARCH=amd64), after the triggered building on amd64
HUB/{image_name}: {image_tag}
HUB/{image_name}: {image_tag}-ARCH
3. Pushing the arm64 images to HUB:(now ARCH=arm64), after the triggered building on arm64
HUB/{image_name}: {image_tag}-ARCH
4. Check the availability of both arm64 and amd64 for
HUB/{image_name}: {image_tag}-ARCH
timely(or trigged?) for all Istio images;
5. Push fat manifest for 
HUB/{image_name}: {image_tag}-amd64
HUB/{image_name}: {image_tag}-arm64
as HUB/{image_name}: {image_tag} to enable multi-arch images
if the check passed

Here we may need to set/get what {image_tag} to check with the version progress of Istio.

**Comment by morlay** [comment]
now ci setup blocked by https://github.com/kubernetes/test-infra/pull/24783

**Comment by morlay** [comment]
@TrevorTaoARM 

Before https://github.com/kubernetes/test-infra support arm64, 
We have to use github actions with self-hosted to build it.

* To use submodule to trace a special branch https://github.com/querycap/istio/blob/release-1.13/.gitmodules
* Then build it when branch changes 
    * https://github.com/querycap/istio/blob/release-1.13/.github/workflows/build-tools.yml
    * https://github.com/querycap/istio/blob/release-1.13/Makefile#L33-L35
    
Then we could got the build-tools for arm64 only for each `release-*`
https://hub.docker.com/r/querycapistio/build-tools/tags?page=1

**Comment by howardjohn** [comment]
Envoy takes like 1 hour to build on 64 core native, with emulation it would take a LONG time.

For build-tools specifically its plausible but looking at >3 hrs last I tried

**Comment by diranged** [comment]
Oh. Yes, that's a fair challenge then... Ouch.

**Comment by zirain** [comment]
> Envoy takes like 1 hour to build on 64 core native, with emulation it would take a LONG time.
> 
> For build-tools specifically its plausible but looking at >3 hrs last I tried

in my env, arm64 take twice more time than x86

**Comment by morlay** [comment]
related https://github.com/istio/test-infra/issues/3413

**Comment by morlay** [comment]
> > Envoy takes like 1 hour to build on 64 core native, with emulation it would take a LONG time.
> > For build-tools specifically its plausible but looking at >3 hrs last I tried
> 
> in my env, arm64 take twice more time than x86

Which size of your device? 

for me, on 12 core arm64, about one hour more or less. (sometimes with cache, it will be faster.)
<img width="319" alt="image" src="https://user-images.githubusercontent.com/1667873/155440573-7241dad8-1a3b-4ae7-a846-ab252539ff3e.png">

**Comment by zirain** [comment]
> > Envoy takes like 1 hour to build on 64 core native, with emulation it would take a LONG time.
> > For build-tools specifically its plausible but looking at >3 hrs last I tried
> 
> in my env, arm64 take twice more time than x86

same size with x86 machine

**Comment by TrevorTaoARM** [comment]
I tested with 40 arm cores(A72 based, Marvell tx2), it takes about 45mins, I think for N1(next generation after A72, for which I think it is what AWS Graviton2 used) based arm cores, it may take less time.
For arm server, it ususually has more cores than those of x86 server, e.g., for Ampere Altra(N1 based), it has 80 physical cores per CPU.

**Comment by infa-ddeore** [comment]
https://github.com/querycap/istio build on c6g.4xlarge Graviton (16 cpu, 32g mem) takes around ~45-50 mins

**Comment by TrevorTaoARM** [comment]
I saw a requirement asking for Istio Arm CI issue in our combined working group meeting minutes, is there any progress here?
https://docs.google.com/document/d/1wsa06GGiq1LEGwhkiPP0FKIZJqdAiue-VeBonWAzAyk/edit#

**Comment by stevenctl** [comment]
@howardjohn Is there someone committed to working on this? I imagine it helps to be a Googler when messing with our Prow cluster, but what can someone from the community do to move this forward.

**Comment by TrevorTaoARM** [comment]
> 2 parts - software and hardware.
> 
> Someone from AWS team is working on getting ARM hardware for us to plug into CI. Not much anyone can do their other than the prow infra admins and the AWS team.
> 
> On the otherside, we need to get our tooling/ci config/etc ready. Its pretty close - the build tools _can_ be built with arm64 today. All we really need, AFAIK, is some job definitions/scripts that actually build them and them combine them into a single multi-arch image.

@howardjohn Thanks.
It's great to learn we are considering the Arm CI. Is there any timeline that the ARM hardware would be available?
How about our next step plan to enable the Arm related jobs? Anything we can do here.

**Comment by howardjohn** [comment]
> It's great to learn we are considering the Arm CI. Is there any timeline that the ARM hardware would be available?

Not sure. Haven't heard from AWS team in a couple weeks.

> How about our next step plan to enable the Arm related jobs? Anything we can do here.

Once we have the software + hardware ready we can turn them on immediately. So we need both of those.

**Comment by howardjohn** [comment]
So specifically we need all the CI tooling to actually build multiarch images on different nodes and join them into a single image

**Comment by TrevorTaoARM** [comment]
@howardjohn Thank you
If you have further progress on the arm HW/SW side, please update this issue to inform us.

**Comment by zioproto** [comment]
Hey @howardjohn , I found this github issue trying to run the linter for a [istio.io docs contribution](https://github.com/istio/istio.io/pull/11539) from my Apple Silicon Mac:

```
% make lint
Unable to find image 'gcr.io/istio-testing/build-tools:release-1.14-f77fd7ec70158fe5e3af4869d3e0c2df152f4c94' locally
release-1.14-f77fd7ec70158fe5e3af4869d3e0c2df152f4c94: Pulling from istio-testing/build-tools
docker: no matching manifest for linux/arm64/v8 in the manifest list entries.
```
I can pull `gcr.io/istio-testing/build-tools:latest` for arm64 but not that specific release tag that the linter wants.

Please do you know if there is any workaround to run the linter on arm64 ? I can't figure out where to patch the Makefile to use a newer version of that build-tool image that supports arm64. 

Thanks folks.

**Comment by howardjohn** [comment]
No versions support arm64 yet (coming soon)

**Comment by howardjohn** [comment]
This has launched now! See https://github.com/istio/istio/issues/26652#issuecomment-1205368297. That comment focuses on the official istio release, but build-tools is multi-arch now as well. Thanks everyone
